File: android/app/src/main/java/com/dontworry/app/domain/search/SynonymExpander.kt
package com.dontworry.app.domain.search

import com.dontworry.app.domain.model.SynonymGroup

class SynonymExpander(groups: List<SynonymGroup>) {
    // Deterministic conflict handling: if a term appears in multiple groups,
    // keep the canonical form that is lexicographically smallest.
    private val canonicalByTerm: Map<String, String> = buildMap {
        groups.forEach { group ->
            val canonical = group.canonical
            group.terms.forEach { term ->
                val current = this[term]
                if (current == null || canonical < current) {
                    put(term, canonical)
                }
            }
        }
    }

    fun expand(queryTerms: List<String>): List<String> {
        if (queryTerms.isEmpty()) return emptyList()
        return queryTerms.map { term -> canonicalByTerm[term] ?: term }
    }
}

File: ios/DontWorryiOS/Domain/SynonymExpander.swift
import Foundation

struct SynonymExpander {
    private let canonicalByTerm: [String: String]

    init(groups: [SynonymGroup]) {
        var map: [String: String] = [:]
        for group in groups {
            let canonical = group.canonical
            for term in group.terms {
                if let current = map[term] {
                    if canonical < current {
                        map[term] = canonical
                    }
                } else {
                    map[term] = canonical
                }
            }
        }
        canonicalByTerm = map
    }

    func expand(queryTerms: [String]) -> [String] {
        guard !queryTerms.isEmpty else { return [] }
        return queryTerms.map { canonicalByTerm[$0] ?? $0 }
    }
}
